src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs

```diff
@@ -88,6 +90,197 @@ public void TestAsyncOutputStream()
+
+                // Wait parent signal to produce number 2
+                // Generate output 2 and signal parent
+                Assert.True(await WaitPipeSignal(pipeRead, WaitInMS), "Missing parent signal to produce number 2");
+                Console.WriteLine(2);
+                await pipeWrite.WriteAsync(new byte[1], 0, 1);
+
+                return SuccessExitCode;
+            }
+        }
+
+        [Fact]
+        [SkipOnTargetFramework(TargetFrameworkMonikers.Uap, "Pipe doesn't work well on UAP")]
+        async public Task TestAsyncOutputStream_BeginCancelBeginOutputRead()
+        {
+            using (AnonymousPipeServerStream pipeWrite = new AnonymousPipeServerStream(PipeDirection.Out, HandleInheritability.Inheritable))
+            using (AnonymousPipeServerStream pipeRead = new AnonymousPipeServerStream(PipeDirection.In, HandleInheritability.Inheritable))
+            {
+                using (Process p = CreateProcessPortable(TestAsyncOutputStream_BeginCancelBeinOutputRead_RemotelyInvokable, $"{pipeWrite.GetClientHandleAsString()} {pipeRead.GetClientHandleAsString()}"))
+                {
+                    var dataReceived = new List<int>();
+                    var dataArrivedEvent = new AutoResetEvent(false);
+
+                    p.StartInfo.RedirectStandardOutput = true;
+                    p.OutputDataReceived += (s, e) =>
+                    {
+                        if (e.Data != null)
+                        {
+                            dataReceived.Add(int.Parse(e.Data));
+                        }
+                        dataArrivedEvent.Set();
+                    };
+
+                    // Start child process
+                    p.Start();
+
+                    pipeWrite.DisposeLocalCopyOfClientHandle();
+                    pipeRead.DisposeLocalCopyOfClientHandle();
+
+                    // Wait child process start
+                    Assert.True(await WaitPipeSignal(pipeRead, WaitInMS), "Child process not started");
+
+                    //Start listening and produce output 1
+                    p.BeginOutputReadLine();
+                    await pipeWrite.WriteAsync(new byte[1], 0, 1);
+
+                    // Wait child signal produce number 1
+                    Assert.True(await WaitPipeSignal(pipeRead, WaitInMS), "Missing child signal for value 1");
+                    Assert.True(dataArrivedEvent.WaitOne(WaitInMS), "Value 1 not received");
+
+                    // Stop listen
+                    p.CancelOutputRead();
+
+                    // clean signal list
+                    dataReceived.Clear();
+
+                    // Produce new value
+                    await pipeWrite.WriteAsync(new byte[1], 0, 1);
+
+                    // Wait child signal produce number 2
+                    Assert.True(await WaitPipeSignal(pipeRead, WaitInMS), "Missing child signal for value 2");
+
+                    // Stop listen
+                    p.BeginOutputReadLine();
```

Comment:
@stephentoub my doubt is here(the reason for wait expiring of dataArrivedEvent), if child produce value and signal but there is a delay it's possible that after begin we receive 2 values and not only 1 because we receive values in async way. I would add ` Assert.False(dataArrivedEvent.WaitOne(WaitInMS), "We don't expect data come in");` before `BeginOutputReadLine` to be sure, I think we cannot have determinism because we don't know how much we need to "elaborate"(skip in case of Cancel state) data coming from child, the best we can is wait for some time(dataArrivedEvent event in this case), tell me your idea.

Reply:
Why does it matter how much we need to skip? The other test validates the correct behavior around not getting data while canceled, right? So this one then just needs to validate that we receive the data after starting up again; it doesn't matter if we also receive data during the perceived cancellation window.

Reply:
> Why does it matter how much we need to skip? The other test validates the correct behavior around not getting data while canceled, right? 

Yes

> So this one then just needs to validate that we receive the data after starting up again; it doesn't matter if we also receive data during the perceived cancellation window.

OK, I let me rephrase your idea, this second test should only verify that if I call `p.CancelOutputRead();` and after `p.BeginOutputReadLine();` after second one I MUST receive some produced data(wait child exit), because first test verify that after `p.CancelOutputRead();` we don't receive data anymore, produce data between cancel/begin doens't make a lot of sense and doesn't have deterministic behaviour, thank's for your perseverance :relaxed:.

Reply:
> after second one I MUST receive some produced data

Yes.  More specifically, you must receive all of the data that you know the child produced after being told to send more:
- Parent starts reading, then signals for the child to data.
- Child sends elements 1, 2, and 3, then waits.
- Parent reads elements 1, 2, 3, verifies it got them, then cancels, then signals to the child that it canceled.
- Child receives cancellation signal and sends next three elements, 4, 5, and 6, then waits.
- Parent immediately (or with a very short delay, doesn't really matter) re-starts reading, and signals to child.
- Child receives signal and sends next three elements, 7, 8, and 9, the exits.
- Parent reads remainder of input, and validates that it at least received 7, 8, and 9 (it might also receive 6, or 5 and 6, or 4 and 5 and 6... that's fine; the other test validated that it doesn't receive input while canceled).

Reply:
> (it might also receive 6, or 5 and 6, or 4 and 5 and 6... that's fine; the other test validated that it doesn't receive input while canceled).

My concern was here...I rework a test with your precise steps, thank's a lot!

Reply:
@stephentoub  did some change, now I'm using BlockingCollection Set/Wait are flaky(very fast set I'll lose some signal).